Add support for painting document markers using system colors

pxlcoder opened the pull request:
#### 68ae85e5b18813ea881d0f18312e848a9389366e
<pre>
Add support for painting document markers using system colors
<a href="https://bugs.webkit.org/show_bug.cgi?id=252791">https://bugs.webkit.org/show_bug.cgi?id=252791</a>
rdar://105796291

Reviewed by Wenson Hsieh and Megan Gardner.

Use system colors for painting document markers when possible. To support this
functionality, new `RenderTheme` methods are added to control platform-specific
marker colors and the color cache is expanded for the new colors. This approach
matches the implementation of existing system colors.

* Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::spellingMarkerColor const):
(WebCore::RenderTheme::platformSpellingMarkerColor const):
(WebCore::RenderTheme::dictationAlternativesMarkerColor const):
(WebCore::RenderTheme::platformDictationAlternativesMarkerColor const):
(WebCore::RenderTheme::autocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::platformAutocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::grammarMarkerColor const):
(WebCore::RenderTheme::platformGrammarMarkerColor const):
(WebCore::RenderTheme::documentMarkerLineColor const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::platformSpellingMarkerColor const):
(WebCore::RenderThemeCocoa::platformDictationAlternativesMarkerColor const):
(WebCore::RenderThemeCocoa::platformAutocorrectionReplacementMarkerColor const):
(WebCore::RenderThemeCocoa::platformGrammarMarkerColor const):
(WebCore::grammarColor): Deleted.
(WebCore::RenderThemeCocoa::documentMarkerLineColor const): Deleted.
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::usePlatformColorForAutocorrectionReplacementMarker):
(WebCore::RenderThemeMac::platformAutocorrectionReplacementMarkerColor const):

Canonical link: <a href="https://commits.webkit.org/260779@main">https://commits.webkit.org/260779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1173f0ede85d675eb9d274b187362a007abb43b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18504 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/42159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19984 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101656 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115128 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/42159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/42159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84834 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11263 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/42159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/42159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13645 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4075 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
```diff
@@ -43,6 +43,12 @@ extern NSString *NSTextCheckingSuppressInitialCapitalizationKey;
+#endif // USE(APPLE_INTERNAL_SDK)
+
+#if HAVE(NSSPELLCHECKER_CORRECTION_INDICATOR_UNDERLINE_COLOR)
+@interface NSSpellChecker (Staging_105286196)
```
If you haven't, can you file a bug for removing this staging code?

pxlcoder replied:
Filed a bug and added a comment.

megangardner commented on Source/WebCore/rendering/RenderThemeCocoa.mm:
```diff
@@ -243,8 +243,33 @@ static inline FontSelectionValue cssWeightOfSystemFont(CTFontRef font)
+Color RenderThemeCocoa::platformDictationAlternativesMarkerColor(OptionSet<StyleColorOptions> options) const
+{
+    auto useDarkMode = options.contains(StyleColorOptions::UseDarkAppearance);
+    return useDarkMode ? SRGBA<uint8_t> { 40, 145, 255, 217 } : SRGBA<uint8_t> { 0, 122, 255, 191 };
```
Is there a way we can factor out all the same special numbers so that it's more clear that they're all the same thing? I'm also assuming there's no way to get these from the system or elsewhere and they need to be manually defined.

pxlcoder replied:
I was doing some digging and it appears that there are system colors that we can use for these!

Though I will need to audit them individually to compare behavior. Since this hardcoding is not new I have filed https://bugs.webkit.org/show_bug.cgi?id=252868 to address this separately.

For now, I have updated the PR to unify the dictation alternative paths, and share the same color method, reducing the amount of duplication.

pxlcoder committed:
Add support for painting document markers using system colors
https://bugs.webkit.org/show_bug.cgi?id=252791
rdar://105796291

Reviewed by Wenson Hsieh and Megan Gardner.

Use system colors for painting document markers when possible. To support this
functionality, new `RenderTheme` methods are added to control platform-specific
marker colors and the color cache is expanded for the new colors. This approach
matches the implementation of existing system colors.

* Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::spellingMarkerColor const):
(WebCore::RenderTheme::platformSpellingMarkerColor const):
(WebCore::RenderTheme::dictationAlternativesMarkerColor const):
(WebCore::RenderTheme::platformDictationAlternativesMarkerColor const):
(WebCore::RenderTheme::autocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::platformAutocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::grammarMarkerColor const):
(WebCore::RenderTheme::platformGrammarMarkerColor const):
(WebCore::RenderTheme::documentMarkerLineColor const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::platformSpellingMarkerColor const):
(WebCore::RenderThemeCocoa::platformDictationAlternativesMarkerColor const):
(WebCore::RenderThemeCocoa::platformAutocorrectionReplacementMarkerColor const):
(WebCore::RenderThemeCocoa::platformGrammarMarkerColor const):
(WebCore::grammarColor): Deleted.
(WebCore::RenderThemeCocoa::documentMarkerLineColor const): Deleted.
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::usePlatformColorForAutocorrectionReplacementMarker):
(WebCore::RenderThemeMac::platformAutocorrectionReplacementMarkerColor const):

Canonical link: https://commits.webkit.org/260779@main
